src/datacollector/DataCollectionManager.cs

```diff
@@ -0,0 +1,665 @@
+        /// Gets cache of data collectors associated with the run.
+        /// </summary>
+        internal Dictionary<Type, DataCollectorWrapper> RunDataCollectors { get; private set; }
+
+        /// <inheritdoc/>
+        public IDictionary<string, string> InitializeDataCollectors(string settingsXml)
+        {
+            ValidateArg.NotNull(settingsXml, "settingsXml");
+
+            var sessionId = new SessionId(Guid.NewGuid());
+            var dataCollectionContext = new DataCollectionContext(sessionId);
+            this.dataCollectionEnvironmentContext = DataCollectionEnvironmentContext.CreateForLocalEnvironment(dataCollectionContext);
+
+            this.attachmentManager.Initialize(sessionId, sourceDirectory, this.messageSink);
+
+            var executionEnvironmentVariables = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+
+            var runConfiguration = XmlRunSettingsUtilities.GetRunConfigurationNode(settingsXml);
+            sourceDirectory = RunSettingsUtilities.GetTestResultsDirectory(runConfiguration);
+
+            var dataCollectionRunSettings = XmlRunSettingsUtilities.GetDataCollectionRunSettings(settingsXml);
+
+            this.isDataCollectionEnabled = dataCollectionRunSettings.IsCollectionEnabled;
+
+            // If dataCollectionRunSettings is null, that means datacollectors are not configured.
+            if (dataCollectionRunSettings == null || !dataCollectionRunSettings.IsCollectionEnabled)
+            {
+                return executionEnvironmentVariables;
+            }
+
+            // Get settings for each data collector, load and initialize the data collectors.
+            var enabledDataCollectorsSettings = this.GetDataCollectorsEnabledForRun(dataCollectionRunSettings);
+            if (enabledDataCollectorsSettings == null || enabledDataCollectorsSettings.Count == 0)
+            {
+                return executionEnvironmentVariables;
+            }
+
+            foreach (var dataCollectorSettings in enabledDataCollectorsSettings)
+            {
+                this.LoadAndInitialize(dataCollectorSettings);
+            }
+
+            // Once all data collectors have been initialized, query for environment variables
+            bool unloadedAnyCollector;
+            var dataCollectorEnvironmentVariables = this.GetEnvironmentVariables(out unloadedAnyCollector);
+
+            foreach (var variable in dataCollectorEnvironmentVariables.Values)
+            {
+                executionEnvironmentVariables.Add(variable.Name, variable.Value);
+            }
+
+            return executionEnvironmentVariables;
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            this.Dispose(true);
+
+            // Use SupressFinalize in case a subclass
+            // of this type implements a finalizer.
+            GC.SuppressFinalize(this);
+        }
+
+        /// <inheritdoc/>
+        public Collection<AttachmentSet> SessionEnded()
```

Comment:
>SessionEnded [](start = 41, length = 12)

This should not return null. If it does, every consumer has to add an extra null check. May be an empty set if better.